flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/table/UpsertKafkaTableITCase.java

```diff
@@ -0,0 +1,385 @@
+
+import org.apache.flink.api.common.restartstrategy.RestartStrategies;
+import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
+import org.apache.flink.streaming.connectors.kafka.KafkaTestBaseWithFlink;
+import org.apache.flink.table.api.EnvironmentSettings;
+import org.apache.flink.table.api.TableResult;
+import org.apache.flink.table.api.bridge.java.StreamTableEnvironment;
+import org.apache.flink.table.planner.factories.TestValuesTableFactory;
+import org.apache.flink.types.Row;
+import org.apache.flink.types.RowKind;
+
+import org.junit.Before;
+import org.junit.Ignore;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+
+import java.sql.Timestamp;
+import java.time.Duration;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Properties;
+import java.util.stream.Collectors;
+
+import static org.apache.flink.streaming.connectors.kafka.table.KafkaTableITCaseUtils.collectRows;
+import static org.apache.flink.streaming.connectors.kafka.table.KafkaTableITCaseUtils.comparedWithKeyAndOrder;
+import static org.apache.flink.streaming.connectors.kafka.table.KafkaTableITCaseUtils.waitingExpectedResults;
+
+/**
+ * Upsert-kafka IT cases.
+ */
+@RunWith(Parameterized.class)
+public class UpsertKafkaTableITCase extends KafkaTestBaseWithFlink {
+
+	private static final String JSON_FORMAT = "json";
+	private static final String CSV_FORMAT = "csv";
+	private static final String AVRO_FORMAT = "avro";
+
+	@Parameterized.Parameter
+	public String format;
+
+	@Parameterized.Parameters(name = "format = {0}")
+	public static Object[] parameters() {
+		return new Object[]{
+				JSON_FORMAT, CSV_FORMAT, AVRO_FORMAT
+		};
+	}
+
+	protected StreamExecutionEnvironment env;
+	protected StreamTableEnvironment tEnv;
+
+	private static final String USERS_TOPIC = "users";
+	private static final String WORD_COUNT_TOPIC = "word_count";
+
+	@Before
+	public void setup() {
+		env = StreamExecutionEnvironment.getExecutionEnvironment();
+		tEnv = StreamTableEnvironment.create(
+				env,
+				EnvironmentSettings.newInstance()
+						.useBlinkPlanner()
+						.inStreamingMode()
+						.build()
+		);
+		env.getConfig().setRestartStrategy(RestartStrategies.noRestart());
+	}
+
+	@Test
+	public void testWordCount() throws Exception {
+		wordCountToUpsertKafka();
+		wordFreqToUpsertKafka();
+		// ------------- clean up ---------------
+		deleteTestTopic(WORD_COUNT_TOPIC + "_" + format);
+	}
+
+	//TODO: test is unstable.
+	@Ignore
+	public void testViewCount() throws Exception {
+		writeChangelogToUpsertKafka();
+		dimensionJoinAndRegisterData();
+		// ------------- clean up ---------------
+		deleteTestTopic(USERS_TOPIC + "_" + format);
+	}
+
+	private void wordCountToUpsertKafka() throws Exception {
+		final String topic = WORD_COUNT_TOPIC + "_" + format;
+		createTopic(topic);
```

Comment:
I would suggest to move the topic creation out of this method, that would be more clear what is created has been cleaned at last.